Add initial E2E tests for `AnyVolumeDatasource` feature

mkimuram opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add initial E2E tests for `AnyVolumeDatasource` feature

#### Which issue(s) this PR fixes:
Fixes #108663

#### Special notes for your reviewer:
/sig storage
@bswartz 

Depends:
 - [x] https://github.com/kubernetes-csi/lib-volume-populator/issues/16
 - [x] https://github.com/kubernetes-csi/volume-data-source-validator/issues/22
 - [x] https://github.com/kubernetes-csi/lib-volume-populator/issues/15


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mkimuram committed:
Implement initial E2E test for AnyVolumeDataSource

mkimuram commented:
Manually tested by below way (and it needs fix in hello-populator side (I will share details in next comment)).

<details>

0. Deploy kind cluster with AnyVolumeDataSource enabled

- In k/k directory of this PR commit, run below:

```bash
kind build node-image

cat << EOF > /tmp/kind-cluster-config.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  AnyVolumeDataSource: true
nodes:
- role: control-plane
  image: kindest/node:latest
- role: worker
  image: kindest/node:latest
EOF

kind create cluster --config /tmp/kind-cluster-config.yaml
```

1. Prepare volume-data-source-validator container image and load it to cluster (TODO: image should be pushed for CI)

```bash
git clone https://github.com/kubernetes-csi/volume-data-source-validator.git
cd volume-data-source-validator
make container-volume-data-source-validator
kind load docker-image volume-data-source-validator:latest docker.io/library/volume-data-source-validator:latest
```

2. Prepare hello-populator container image and load it to cluster (TODO: image should be pushed for CI)

```bash
cd ..
git clone https://github.com/kubernetes-csi/lib-volume-populator.git
cd lib-volume-populator 
make container-hello-populator
kind load docker-image hello-populator:latest docker.io/library/hello-populator:latest
```

3. Deploy CRDs in the cluster (TODO: needs to be automated for CI)

```bash
kubectl apply -f test/e2e/testing-manifests/any-volume-datasource/populator.storage.k8s.io_volumepopulators.yaml 
kubectl apply -f test/e2e/testing-manifests/any-volume-datasource/hello-populator-crd.yaml
```

(If CRDs for snapshot are missing, also deploy like below)

```bash
kubectl apply -f cluster/addons/volumesnapshots/crd/snapshot.storage.k8s.io_volumesnapshots.yaml
kubectl apply -f cluster/addons/volumesnapshots/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml 
kubectl apply -f cluster/addons/volumesnapshots/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml 
```

4. Run the E2E tests

- In k/k directory of this PR commit, run below:

```bash
make ginkgo
make all WHAT=test/e2e/e2e.test
KUBERNETES_PROVIDER=custom KUBERNETES_CONFORMANCE_TEST=y hack/ginkgo-e2e.sh --ginkgo.focus="\[Driver: csi-hostpath\].*\[Feature:AnyVolumeDataSource\].*"
```

</details>

mkimuram commented:
@bswartz Also, it may be out of scope of this PR, but I found that `VolumePopulator` resource is defined as [namespaced](https://github.com/kubernetes-csi/volume-data-source-validator/blob/master/client/config/crd/populator.storage.k8s.io_volumepopulators.yaml#L18). Shouldn't it be cluster scoped, because users in different namespaces would need to check which controller is used for which resources and cluster admins would need to ensure that one resource is not handled by multiple controllers?

xing-yang commented on test/e2e/testing-manifests/storage-csi/any-volume-datasource/setup-data-source-validator.yaml:
```diff
@@ -0,0 +1,24 @@
+      serviceAccount: volume-data-source-validator
+      containers:
+        - name: volume-data-source-validator
+          image: volume-data-source-validator:latest
```
This should be a released image.

mkimuram replied:
I found that `k8s.gcr.io/sig-storage/volume-data-source-validator:v0.1.0` and `k8s.gcr.io/sig-storage/hello-populator:v0.1.0` are already published.  I've updated to use these images and opened issues to update the doc for these repo (and added references for the issues in the first comment).

xing-yang commented on test/e2e/testing-manifests/storage-csi/any-volume-datasource/populator.storage.k8s.io_volumepopulators.yaml:
```diff
@@ -0,0 +1,54 @@
+
```
CRDs should not be installed here.  There are a couple of places where CRDs will be installed:
For in-tree, see how CRDs are installed for VolumeSnapshot CRDs:
https://github.com/kubernetes/kubernetes/pull/80058
@msau42 @mattcary Are there any changes regarding this?

CRDs also need to be installed by csi-release-tools:
https://github.com/kubernetes-csi/csi-release-tools/blob/master/prow.sh#L741

mkimuram replied:
>> For in-tree, see how CRDs are installed for VolumeSnapshot CRDs:
> I will update the code.

Added.

mkimuram replied:
@xing-yang  Should we deploy hello-populator in each test or can we deploy it once on cluster creation?
The current code only deploys volume-data-source-validator on cluster creation and deploys hello-populator in each test (as volume-snapshot-controller is deployed on cluster creation, and each snapshotter is deployed in each test for snapshot tests).

If we agree to deploy hello-populator once on cluster creation, we might be able to go ahead without the ability to deploy hello-populator in any namespaces. Although, controllers in general should be able to be deployed in any namespace and we can deploy hello-populator to different namespace from the test namespace, as existing provisioner tests are doing.

bswartz commented:
> @bswartz As a result of the test by using the way described in my previous comment, I see the below log in hello-populator container. As seen in current [hello-populator code](https://github.com/kubernetes-csi/lib-volume-populator/blob/master/example/hello-populator/main.go#L34), it only handles the PVC in hello name space, so the test running in another namespace fails (I think that most E2E tests run in random namespace to avoid affected by other tests running in parallel, so it would be better to fix it by hello-populator side). Can it be improved?

Hey, thanks for writing these tests! If I understand the issue, the fact that the sample populator has a hardcoded namespace is the source of the problem. What if the namespace was configurable as a command line argument? Would that fix this problem?

bswartz commented:
> @bswartz Also, it may be out of scope of this PR, but I found that `VolumePopulator` resource is defined as [namespaced](https://github.com/kubernetes-csi/volume-data-source-validator/blob/master/client/config/crd/populator.storage.k8s.io_volumepopulators.yaml#L18). Shouldn't it be cluster scoped, because users in different namespaces would need to check which controller is used for which resources and cluster admins would need to ensure that one resource is not handled by multiple controllers?

This is a big mistake. It should be non-namespaced. Let me see if I can figure out what happened here.

mkimuram commented:
> If I understand the issue, the fact that the sample populator has a hardcoded namespace is the source of the problem. What if the namespace was configurable as a command line argument? Would that fix this problem?

@bswartz  Yes. As an example populator for this E2E test, it would be good enough.

landreasyan commented:
> > If I understand the issue, the fact that the sample populator has a hardcoded namespace is the source of the problem. What if the namespace was configurable as a command line argument? Would that fix this problem?
> 
> @bswartz Yes. As an example populator for this E2E test, it would be good enough.

Hi, I don't think changing the populator namespace to the e2e namespace is correct. The hardcoded namespace is for handling the internal workings of the populator, meaning the creation of the extra PVC and Pod for copying the data and it is not the namespace for PVCs that exist in the cluster. The controller monitors all the necessary resources in the cluster and not just the ones in "hello" namespace.  We want to keep the namespace for the populator separately, because the controller assumes that all the resources in the populator namespace are solely for managing the "copying" of the data. 
The issue here is the fact that we created the hello populator in the e2e namespace, so there are no resources that triggered the creation of "hello" namespace that the populator needs. So, we could either create the "hello" namespace or create a separate namespace(from the e2e one) for the populator and pass it as a command line arg to the hello controller.

mkimuram commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -247,7 +249,38 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
+							args := []string{}
+							for _, arg := range container.Args {
+								if strings.HasPrefix(arg, "--namespace=") {
+									args = append(args, fmt.Sprintf("--namespace=%s", f.Namespace.Name))
```
@bswartz @landreasyan I've create a PR for hello-populator to accept `--namespace` argument as  https://github.com/kubernetes-csi/lib-volume-populator/pull/17 and updated this PR to patch the hello-populator-deploy.yaml to add `--namespace` argument. 

If populator should be deployed to different namespace to the test namespace, we can change the code to create another namespace for populator and deploy it to the namespace.

mkimuram replied:
> If populator should be deployed to different namespace to the test namespace, we can change the code to create another namespace for populator and deploy it to the namespace.

The last commit "Create populator in a separate namespace" handles it
(and the current codes can be manually tested by the steps in https://github.com/kubernetes/kubernetes/pull/108664#issuecomment-1067365541).

mkimuram commented:
Note that the current test code can be manually run by below way:

<details>

0. Deploy cluster either by below ways (In my environment, deploying by (a) didn't work and I'm not sure why, though)

In k/k of this PR commit:

(a) By using hack/local-up-cluster.sh
```bash
ENABLE_VOLUME_DATA_SOURCE_VALIDATOR=true FEATURE_GATES="AnyVolumeDataSource=true" hack/local-up-cluster.sh
```

Or

(b) By using kind
```bash
kind build node-image
cat << EOF > /tmp/kind-cluster-config.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  AnyVolumeDataSource: true
nodes:
- role: control-plane
  image: kindest/node:latest
- role: worker
  image: kindest/node:latest
EOF

kind create cluster --config /tmp/kind-cluster-config.yaml

kubectl apply -f cluster/addons/any-volume-datasource/crd/
kubectl apply -f cluster/addons/any-volume-datasource/volume-data-source-validator/
kubectl apply -f cluster/addons/volumesnapshots/crd/
```

1. Build --namespace supported version of hello-populator  (This won't be needed once the hello-populator is updated and published image version is updated)

```bash
git clone https://github.com/mkimuram/lib-volume-populator.git -b issue/16
cd lib-volume-populator 
make container-hello-populator
docker tag hello-populator:latest k8s.gcr.io/sig-storage/hello-populator:v0.1.1 
kind load docker-image k8s.gcr.io/sig-storage/hello-populator:v0.1.1 
```

2. Make the E2E test use the image (This won't be needed once the hello-populator is updated and published image version is updated)
- In k/k of this PR commit
```bash
sed -i 's|k8s.gcr.io/sig-storage/hello-populator:v0.1.0|k8s.gcr.io/sig-storage/hello-populator:v0.1.1|' test/e2e/testing-manifests/storage-csi/any-volume-datasource/hello-populator-deploy.yaml 
```

3. Run the tests
```bash
make ginkgo
make all WHAT=test/e2e/e2e.test
KUBERNETES_PROVIDER=custom KUBERNETES_CONFORMANCE_TEST=y hack/ginkgo-e2e.sh --ginkgo.focus="\[Driver: csi-hostpath\].*\[Feature:AnyVolumeDataSource\].*"
```

</details>

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

This PR should be ready to be tested with alpha features job.

mkimuram commented:
I've created a PR to enable this e2e test in job pull-kubernetes-e2e-gce-alpha-features. Please check https://github.com/kubernetes/test-infra/pull/25709 .

mkimuram commented:
/retest

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

mkimuram commented:
/test all

mkimuram commented:
@xing-yang It seems that the test [fails](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108664/pull-kubernetes-e2e-gce-alpha-features/1506685033644036096/) due to the lack of VolumePopulator CRD [here](https://github.com/kubernetes/kubernetes/pull/108664/files#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R321). But it was intended to be created [here](https://github.com/kubernetes/kubernetes/pull/108664/files#diff-ada96bec4aba13195248d6dd21ee67e333bead2bd898b81a1b089d53bd06b4d8R2902), as the snapshot tests do. Could you help me find what's wrong?

I'm not familiar with how addons are enabled in the gce test job, like who calls config-test.sh and where, I would appreciate it if you could let me know the pointer to such information.

xing-yang commented on hack/local-up-cluster.sh:
```diff
@@ -940,6 +943,18 @@ function start_csi_snapshotter {
+    if [[ "${ENABLE_VOLUME_DATA_SOURCE_VALIDATOR}" = true ]]; then
+        echo "Creating volume data source validator"
+        ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/any-volume-datasource/crd/populator.storage.k8s.io_volumepopulators.yaml"
+        ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/any-volume-datasource/crd/hello-populator-crd.yaml"
```
The Hello CRDs should be installed as part of the test (like what you did before).
Volume Populator CRDs should be installed by the cluster as what you have here.

mkimuram replied:
Thank you for your comment. It seems that we need to check [kube-master-configuration.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/108664/pull-kubernetes-e2e-gce-alpha-features/1506685033644036096/artifacts/e2e-d4b68bf607-cc220-master/kube-master-configuration.log) for the addon's setup logs. And it seems failing.

```
Mar 23 18:07:52.448034 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: error: the server doesn't have a resource type "VolumePopulator"
Mar 23 18:07:52.749176 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: No resources found in default namespace.
Mar 23 18:07:52.750579 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: Wait until volume snapshot RBAC rules are installed
Mar 23 18:07:52.998490 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: NAME                              ROLE                                            AGE
Mar 23 18:07:52.998995 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: volume-snapshot-controller-role   ClusterRole/volume-snapshot-controller-runner   9s
Mar 23 18:07:53.000741 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: Wait until volume snapshot controller is installed
Mar 23 18:07:53.239418 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: Error from server (NotFound): statefulsets.apps "volume-snapshot-controller" not found
Mar 23 18:08:02.766193 e2e-d4b68bf607-cc220-master configure-helper.sh[1288]: error: the server doesn't have a resource type "VolumePopulator"
```

I added `addonmanager.kubernetes.io/mode=Reconcile` labels to all the VolumePopulator related resources, which looks like the biggest difference from snapshot ones, and fixed the check logics. And let me continue debugging.

xing-yang commented:
> @xing-yang It seems that the test [fails](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108664/pull-kubernetes-e2e-gce-alpha-features/1506685033644036096/) due to the lack of VolumePopulator CRD [here](https://github.com/kubernetes/kubernetes/pull/108664/files#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R321). But it was intended to be created [here](https://github.com/kubernetes/kubernetes/pull/108664/files#diff-ada96bec4aba13195248d6dd21ee67e333bead2bd898b81a1b089d53bd06b4d8R2902), as the snapshot tests do. Could you help me find what's wrong?

The AnyVolumeDataSource feature gate is not enabled by default yet.  That's probably why the CRDs are not deployed.  Can you add some messages in the script where you install the CRDs?

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

mkimuram commented:
Now below tests pass:

- `[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`
- `[sig-storage] In-tree Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`
- `[sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io][Serial] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`
- `[sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: block] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`
- `[sig-storage] In-tree Volumes [Driver: rbd][Feature:Volumes][Serial] [Testpattern: Dynamic PV (ntfs)][Feature:Windows] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`

And only below one test with in-tree nfs driver fails:

- `[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`

The error message below seems to say that the expected file isn't found in the volume.

```
{ Failure test/e2e/storage/testsuites/provisioning.go:244
Mar 24 00:24:52.932: Unexpected error:
    <exec.CodeExitError>: {
        Err: {
            s: "error running /go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/kubectl --server=https://35.185.234.210 --kubeconfig=/workspace/.kube/config --namespace=provisioning-7288 exec nfs-client --namespace=provisioning-7288 -- cat /opt/0/example-provisioning-7288.txt:\nCommand stdout:\n\nstderr:\ncat: can't open '/opt/0/example-provisioning-7288.txt': No such file or directory\ncommand terminated with exit code 1\n\nerror:\nexit status 1",
        },
        Code: 1,
    }
    error running /go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/kubectl --server=https://35.185.234.210 --kubeconfig=/workspace/.kube/config --namespace=provisioning-7288 exec nfs-client --namespace=provisioning-7288 -- cat /opt/0/example-provisioning-7288.txt:
    Command stdout:
    
    stderr:
    cat: can't open '/opt/0/example-provisioning-7288.txt': No such file or directory
    command terminated with exit code 1
    
    error:
    exit status 1
occurred
test/e2e/framework/util.go:630}
```

@bswartz Are there any expected reasons that only nfs driver can fail? (RWX volume, in-tree driver, etc ...?)
Also, I'm wondering if this feature can/should work with in-tree drivers in the first place.

mkimuram commented:
@bswartz 

What will be the expected behavior for in-tree drivers when `DataSourceRef` referencing CRD is specified?

Should populator populate the data for in-tree driver?
If not, how should it be avoided?
(According to the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators#summary), it seems "CSI-specific").

I can disable the e2e tests for in-tree driver for now. However, I think that above should be clear before doing it
(we may decide to just wait for in-tree drivers to be deprecated or to just allow them populated without declaring support, but we would need decision).  

Any thoughts? @xing-yang

xing-yang commented:
We probably should add a check similar to this: https://github.com/kubernetes/kubernetes/pull/97086/files#diff-9a553e2f939f7ad1470cfe1cf8a48b56b39c6b34fb7ad820c79f2f43602499bdR1477

bswartz commented:
> @bswartz
> 
> What will be the expected behavior for in-tree drivers when `DataSourceRef` referencing CRD is specified?

I would expect the same behavior that external provisioner has, which is to generate an event on the PVC saying that the data source is not recognized, and then do nothing. If there are in-tree provisioners that don't do this, I'd like to fix them.

mkimuram commented on cluster/addons/any-volume-datasource/crd/populator.storage.k8s.io_volumepopulators.yaml:
```diff
@@ -0,0 +1,56 @@
+
```
Added.

mkimuram commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -242,6 +242,9 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
 	})
 
 	ginkgo.It("should provision storage with any volume data source [Feature:AnyVolumeDataSource]", func() {
+		if len(dInfo.InTreePluginName) != 0 {
```
Disabled this test for in-tree drivers.

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

mkimuram commented:
Now these tests pass in pull-kubernetes-e2e-gce-alpha-features job:
- `[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`
- `[sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io][Serial] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Feature:AnyVolumeDataSource]`

Can this PR be merged?

There are some remaining jobs for the implementation side, but I think that it would be better to add the test first and incrementally improve the codes along with the tests.

bswartz commented:
@mkimuram I wanted to let you know that new images have been published for beta builds of hello-populator and volume-data-source-validator
k8s.gcr.io/sig-storage/hello-populator:v1.0.0
k8s.gcr.io/sig-storage/volume-data-source-validator:v1.0.0

mkimuram commented:
@bswartz Thank you for letting me know. Updated to use v1.0.0 images and also updated VolumePopulator to cluster-scoped. I will squash the commits once this PR is ready to be merged.

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

mkimuram commented:
This PR should be ready to be merged. PTAL @bswartz @xing-yang

bswartz reviewed (approved):
/lgtm

xing-yang commented:
@mkimuram Can you squash your commits and add a release note?

mkimuram commented:
@xing-yang  Sure. Is it OK to just merge them all into one or do you prefer to keep the first 4 commits and merge rest of them into proper ones? (Or maybe keeping the first and the second one, and merge rest of them?)

xing-yang commented:
Either merge all in one commit or put changes in test folder in one commit and the rest of them in another.

mkimuram commented:
Merged all in one commit.

mkimuram commented:
/test pull-kubernetes-e2e-gce-alpha-features

I guess we need to trigger this alpha job, until we add a job like [this](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-storage/sig-storage-gce-config.yaml#L54), even after this feature itself is marked as Beta. I will start working on adding it once this PR is merged (and before test freeze for 1.24).

mkimuram commented:
/retest

xing-yang commented:
@mkimuram Can you add a release note?

xing-yang commented:
/lgtm
/approve

xing-yang commented:
/assign @dims

mkimuram commented:
@xing-yang Sorry. I forgot to update the release note field. Added `NONE`, as this PR just adds tests.
@dims Could you approve this PR?

dims commented:
/assign @BenTheElder

mkimuram commented:
@BenTheElder Could you approve this PR?
@dims I think that we need your approval too for image approval.

dims commented:
@mkimuram it would be good for another hack/ cluster/ reviewer like Ben to do first. and then i can deal with the images/ directory please.

liggitt commented:
we *really* shouldn't be adding new things to `cluster`

liggitt commented:
if a test needs test-only manifests, can we put them somewhere like test/e2e/testing-manifests/...?

mkimuram commented:
@liggitt @BenTheElder  Thank you for your comments.

> I haven't had a chance to look at exactly what this addon does yet, but we have other types of cluster in CI (e.g. kops, kubernetes-sigs/kind, ...) that can probably run this if local-up-cluster.sh can, but not if we're special-casing install.

Actually, I've also added a script to deploy all the same necessary things in [hack/local-up-cluster.sh](https://github.com/kubernetes/kubernetes/pull/108664/commits/844283a35b5a3052542d526dc097690fd396c2c8#diff-c864163457a067ed45f7c856114c77716d1f558b11184df3fb2142992553e1cdR945) (It deploys them, if `ENABLE_VOLUME_DATA_SOURCE_VALIDATOR == true`). Therefore, if you let me know which jobs are using hack/local-up-cluster.sh and a reference to how the variable is passed, I will update this PR to use only hack/local-up-cluster.sh and remove all the file under the cluster directory.

BenTheElder commented:
> Therefore, if you let me know which jobs are using hack/local-up-cluster.sh

There are only one or two, most of CI will be using cluster/ or kubernetes-sigs/kind instead, or perhaps kops or cluster-api.
This script is mostly for local development, and I'm not sure we should be adding addons to local-up-cluster either.

liggitt commented:
> I'm not sure we should be adding addons to local-up-cluster either.

agreed. I'm more looking for tests to bring their own manifests than to bake it into cluster setup

liggitt commented on cluster/addons/any-volume-datasource/crd/hello-populator-crd.yaml:
```diff
@@ -0,0 +1,54 @@
+  labels:
+    addonmanager.kubernetes.io/mode: Reconcile
+  annotations:
+    api-approved.kubernetes.io: unapproved
```
If this is a test, it should use example.com

mkimuram replied:
Fixed as suggested.

BenTheElder commented on cluster/addons/any-volume-datasource/crd/hello-populator-crd.yaml:
```diff
@@ -0,0 +1,54 @@
+  labels:
+    addonmanager.kubernetes.io/mode: Reconcile
+  annotations:
+    api-approved.kubernetes.io: unapproved
```
should we really be installing new unapproved API types?
this seems like it should either be approved or in x-k8s.io

liggitt replied:
that should definitely be updated there... needing to add "unapproved" is a good sign this shouldn't be using this suffix

liggitt replied:
a test-only fixture could use something like `example.com`

mkimuram replied:
Fixing in https://github.com/kubernetes-csi/lib-volume-populator/pull/23 and will update this PR once it is merged.

mkimuram replied:
The change in Hello populator was merged and I'm waiting for the container image to be updated. 
As for the change in this PR, I've tested locally with below change by using locally built image and confirmed that it works. I will update this PR once the container image is updated.
https://github.com/mkimuram/kubernetes/commit/c150d2564fd369a3394805cc08a58d76f36cd76c

mkimuram replied:
The updated image has been published as v1.0.1, so I've updated this PR as I [mentioned](https://github.com/kubernetes/kubernetes/pull/108664#discussion_r841095990). So, this PR should be ready to be merged.

/hold cancel

BenTheElder commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -239,6 +243,156 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
+			f.DeleteNamespace(popNamespace.Name)
+		}()
+
+		ginkgo.By("Deploying hello-populator")
```
so is this test dependent on the hello type which it does not install and is not in core?

generally not a fan of adding e2e tests that are never going to be portable.

if there are core kubernetes features / code changes / ... related to this, we should come up with a portable way to test them. if it's entirely out-of-tree, we should test it out-of-tree.

mkimuram replied:
The intention of the test is not to test the example populator itself, but to test that one of the populators actually works. But I agree with your point on portability, so let me think about it if there is a good way.

mkimuram commented:
>> I'm not sure we should be adding addons to local-up-cluster either.

> agreed. I'm more looking for tests to bring their own manifests than to bake it into cluster setup

We might be able to deploy them inside the tests, if the tests don't run in parallel by adding "[Serial]" tag. Let me try changing like such a way.

mkimuram commented:
Moved the validator deploying [codes](https://github.com/kubernetes/kubernetes/pull/108664/commits/0dda878c5566254b86baa2245ecd27e9175b6f35#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R268-R273) and example CRD deploying [codes](https://github.com/kubernetes/kubernetes/pull/108664/commits/0dda878c5566254b86baa2245ecd27e9175b6f35#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R294) inside the tests. Now, this PR doesn't rely on cluster/ and hack/ .

The tests need `[Serial]` to run in parallel, but adding it will make no job run the tests, so let me:
/hold
/test pull-kubernetes-e2e-gce-alpha-features

@dims @BenTheElder Please check if this looks good. If so, I will work on updating test-infra side in https://github.com/kubernetes/test-infra/pull/25823.

mkimuram commented:
Or should we just add `[Serial]` and remove `[Feature:AnyVolumeDataSource]`, because there will be no dependency once this feature become Beta now? WDYT? @xing-yang

xing-yang commented:
> Or should we just add [Serial] and remove [Feature:AnyVolumeDataSource], because there will be no dependency once this feature become Beta now? WDYT? @xing-yang

@mkimuram, I think we should add a separate job similar to pull-kubernetes-e2e-gce-storage-snapshot for volume populator.

mkimuram commented:
>>Or should we just add [Serial] and remove [Feature:AnyVolumeDataSource], because there will be no dependency once this feature become Beta now? WDYT? @xing-yang

> @mkimuram, I think we should add a separate job similar to pull-kubernetes-e2e-gce-storage-snapshot for volume populator.

@xing-yang Thank you for your comment.

Current codes don't do any special setups in cluster/ and hack/ for the tests , so the tests should work well on any environments. Actually, by adding [only `[Serial]` tag](https://github.com/kubernetes/kubernetes/pull/108664/commits/e953bff0f292c716e7720b96835181f72d4b05a8#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R246), I confirmed that [pull-kubernetes-e2e-gce-csi-serial job](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108664/pull-kubernetes-e2e-gce-csi-serial/1510069782949400576/) runs below two tests:

- `Kubernetes e2e suite: [sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io][Serial] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Serial]`
- `Kubernetes e2e suite: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with any volume data source [Serial]`

Should we still need to add a new separate job like pull-kubernetes-e2e-gce-storage-populator for volume populator?


I mean that (a) works, but should we choose (b), just to be sure? @msau42

(a) add only `[Serial]` tag to the tests
   -> jobs like pull-kubernetes-e2e-gce-csi-serial, which run `[Serial]` tests, run the tests
(b) add both `[Serial]` and `[Feature:AnyVolumeDataSource]` tags to the tests and add a new pull-kubernetes-e2e-gce-storage-populator job to run these tags
   -> only pull-kubernetes-e2e-gce-storage-populator job runs the tests

xing-yang commented:
/lgtm
I'm fine to get this merged first.

BenTheElder commented:
Does it require a non-default feature gate? If not and the tests can install the remaining dependency themselves it probably doesn't need a feature tag.

BenTheElder commented:
I am out monday-friday this week 4th- and only got a quick look tonight, if this is aiming for test freeze then please reassign, I think multiple other people on this thread can do the remaining approval

Otherwise, happy to follow up next week.

mkimuram commented:
@BenTheElder

> Does it require a non-default feature gate? If not and the tests can install the remaining dependency themselves it probably doesn't need a feature tag.

No. AnyVolumeDatasource feature, which is the only feature that the test depends, has already been marked as beta. So, the test can now work without other dependencies, as everything needed is deployed inside each test case.


> I am out monday-friday this week 4th- and only got a quick look tonight, if this is aiming for test freeze then please reassign, I think multiple other people on this thread can do the remaining approval
>
> Otherwise, happy to follow up next week.

Thank you for your review. It's very helpful. I'm aiming for the test freeze, but other people already assigned will be able to approve, as you mentioned. Enjoy your holiday!

mkimuram commented:
@dims Could you review and approve image/ part of this PR? 

This PR adds two images, "volume-data-source-validator" and "hello-populator", to [expectedImages](https://github.com/kubernetes/kubernetes/pull/108664/commits/cd634b59ac69145860712d56fbb976dedb940b1e#diff-76f8e0d202004c8493f25531d973581d97a42b38ecf70120548dd683621ad8b0R55-R56) for CSI tests.
They are used for testing AnyVolumeDataSource feature (please also see [here](https://kubernetes.io/blog/2021/08/30/volume-populators-redesigned/) and [here](https://kubernetes.io/blog/2021/08/30/volume-populators-redesigned/)). As we [decided](https://github.com/kubernetes/kubernetes/pull/108664#issuecomment-1086266687) not to deploy them in cluster/ or hack/, we need to deploy them inside each e2e test by using CreateFromManifests [here](https://github.com/kubernetes/kubernetes/pull/108664/commits/cd634b59ac69145860712d56fbb976dedb940b1e#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R269-R274) and [here](https://github.com/kubernetes/kubernetes/pull/108664/commits/cd634b59ac69145860712d56fbb976dedb940b1e#diff-c0b41d6b4f963df4d44f35004220c427e991fa0606ec6bdc3ffbeef38e27bba6R293-R297), so the images used there needed to be added to expectedImages.

dims commented:
@mkimuram looks way better :) thanks.

reviewed changes in `test/utils/image`

/approve
/lgtm

mkimuram commented:
/retest

xing-yang commented:
/triage accepted

xing-yang commented:
/priority important-soon

xing-yang commented:
/milestone v1.24
